// ngx6-angular-breadcrumbs/esm5/src/component/mc-breadcrumbs.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { McBreadcrumbsService } from '../service/mc-breadcrumbs.service';
import { Component } from '@angular/core';
var McBreadcrumbsComponent = /** @class */ (function () {
    function McBreadcrumbsComponent(service) {
        this.service = service;
        this.subscriptions = new Array();
    }
    /**
     * @return {?}
     */
    McBreadcrumbsComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        /** @type {?} */
        var s = this.service.crumbs$.subscribe(function (x) {
            _this.crumbs = x;
        });
    };
    /**
     * @return {?}
     */
    McBreadcrumbsComponent.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        this.subscriptions.forEach(function (x) { return x.unsubscribe(); });
    };
    McBreadcrumbsComponent.decorators = [
        { type: Component, args: [{
                    // tslint:disable-next-line:component-selector
                    selector: 'mc-breadcrumbs',
                    template: "<ol *ngIf=\"crumbs.length\" class=\"breadcrumb\">\n  <li *ngFor=\"let crumb of crumbs; let last = last\" [ngClass]=\"{ 'active': last }\" class=\"breadcrumb-item\">\n    <a *ngIf=\"!last\" [routerLink]=\"crumb.path\">{{ crumb.text }}</a>\n    <span *ngIf=\"last\">{{ crumb.text }}</span>\n  </li>\n</ol>\n"
                },] },
    ];
    /** @nocollapse */
    McBreadcrumbsComponent.ctorParameters = function () { return [
        { type: McBreadcrumbsService }
    ]; };
    return McBreadcrumbsComponent;
}());
export { McBreadcrumbsComponent };
if (false) {
    /** @type {?} */
    McBreadcrumbsComponent.prototype.crumbs;
    /** @type {?} */
    McBreadcrumbsComponent.prototype.subscriptions;
    /** @type {?} */
    McBreadcrumbsComponent.prototype.service;
}
