// ngx6-angular-breadcrumbs/esm2015/src/mc-breadcrumbs.shared.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { from, of, Observable } from 'rxjs';
import * as template from 'lodash.template';
import * as templateSettings from 'lodash.templatesettings';
/** @type {?} */
const _ = {
    template: template,
    templateSettings: templateSettings
};
_.templateSettings.interpolate = /{{([\s\S]+?)}}/g;
/**
 * @record
 */
export function IBreadcrumb() { }
/** @type {?} */
IBreadcrumb.prototype.text;
/** @type {?} */
IBreadcrumb.prototype.path;
/**
 * @param {?} module
 * @return {?}
 */
function r(module) {
    return require(`${name}`);
}
/**
 * @param {?} template
 * @param {?} binding
 * @return {?}
 */
export function stringFormat(template, binding) {
    /** @type {?} */
    const compiled = _.template(template);
    return compiled(binding);
}
/**
 * @param {?} value
 * @return {?}
 */
export function isPromise(value) {
    return value && typeof value.then === 'function';
}
/**
 * @template T
 * @param {?} value
 * @return {?}
 */
export function wrapIntoObservable(value) {
    if (value instanceof Observable) {
        return value;
    }
    if (isPromise(value)) {
        return from(Promise.resolve(value));
    }
    return of(/** @type {?} */ (value));
}
