// ngx6-angular-breadcrumbs/esm5/src/mc-breadcrumbs.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { RouterModule } from '@angular/router';
import { CommonModule } from '@angular/common';
import { McBreadcrumbsService } from './service/mc-breadcrumbs.service';
import { NgModule } from '@angular/core';
import { McBreadcrumbsComponent } from './component/mc-breadcrumbs.component';
import { McBreadcrumbsConfig } from './service/mc-breadcrumbs.config';
var McBreadcrumbsModule = /** @class */ (function () {
    function McBreadcrumbsModule() {
    }
    /**
     * @return {?}
     */
    McBreadcrumbsModule.forRoot = /**
     * @return {?}
     */
    function () {
        return {
            ngModule: McBreadcrumbsModule,
            providers: [McBreadcrumbsService, McBreadcrumbsConfig]
        };
    };
    McBreadcrumbsModule.decorators = [
        { type: NgModule, args: [{
                    imports: [CommonModule, RouterModule],
                    declarations: [McBreadcrumbsComponent],
                    exports: [McBreadcrumbsComponent]
                },] },
    ];
    return McBreadcrumbsModule;
}());
export { McBreadcrumbsModule };
