// ngx6-angular-breadcrumbs/esm2015/src/component/mc-breadcrumbs.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { McBreadcrumbsService } from '../service/mc-breadcrumbs.service';
import { Component } from '@angular/core';
export class McBreadcrumbsComponent {
    /**
     * @param {?} service
     */
    constructor(service) {
        this.service = service;
        this.subscriptions = new Array();
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        /** @type {?} */
        const s = this.service.crumbs$.subscribe(x => {
            this.crumbs = x;
        });
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        this.subscriptions.forEach(x => x.unsubscribe());
    }
}
McBreadcrumbsComponent.decorators = [
    { type: Component, args: [{
                // tslint:disable-next-line:component-selector
                selector: 'mc-breadcrumbs',
                template: `<ol *ngIf="crumbs.length" class="breadcrumb">
  <li *ngFor="let crumb of crumbs; let last = last" [ngClass]="{ 'active': last }" class="breadcrumb-item">
    <a *ngIf="!last" [routerLink]="crumb.path">{{ crumb.text }}</a>
    <span *ngIf="last">{{ crumb.text }}</span>
  </li>
</ol>
`
            },] },
];
/** @nocollapse */
McBreadcrumbsComponent.ctorParameters = () => [
    { type: McBreadcrumbsService }
];
if (false) {
    /** @type {?} */
    McBreadcrumbsComponent.prototype.crumbs;
    /** @type {?} */
    McBreadcrumbsComponent.prototype.subscriptions;
    /** @type {?} */
    McBreadcrumbsComponent.prototype.service;
}
