// ngx6-angular-breadcrumbs/fesm5/ngx-breadcrumbs.js
import { from, of, Observable, BehaviorSubject } from 'rxjs';
import * as template from 'lodash.template';
import * as templateSettings from 'lodash.templatesettings';
import { Injectable, Injector, Component, NgModule } from '@angular/core';
import { NavigationEnd, Router, RouterModule } from '@angular/router';
import { flatMap, distinct, toArray, concat, first, filter } from 'rxjs/operators';
import { CommonModule } from '@angular/common';

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
/** @type {?} */
var _ = {
    template: template,
    templateSettings: templateSettings
};
_.templateSettings.interpolate = /{{([\s\S]+?)}}/g;
/**
 * @param {?} template
 * @param {?} binding
 * @return {?}
 */
function stringFormat(template$$1, binding) {
    /** @type {?} */
    var compiled = _.template(template$$1);
    return compiled(binding);
}
/**
 * @param {?} value
 * @return {?}
 */
function isPromise(value) {
    return value && typeof value.then === 'function';
}
/**
 * @template T
 * @param {?} value
 * @return {?}
 */
function wrapIntoObservable(value) {
    if (value instanceof Observable) {
        return value;
    }
    if (isPromise(value)) {
        return from(Promise.resolve(value));
    }
    return of(/** @type {?} */ (value));
}

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
var McBreadcrumbsConfig = /** @class */ (function () {
    function McBreadcrumbsConfig() {
    }
    McBreadcrumbsConfig.decorators = [
        { type: Injectable },
    ];
    return McBreadcrumbsConfig;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
var McBreadcrumbsResolver = /** @class */ (function () {
    function McBreadcrumbsResolver() {
    }
    /**
     * @param {?} route
     * @param {?} state
     * @return {?}
     */
    McBreadcrumbsResolver.prototype.resolve = /**
     * @param {?} route
     * @param {?} state
     * @return {?}
     */
    function (route, state) {
        /** @type {?} */
        var data = route.routeConfig.data;
        /** @type {?} */
        var path = this.getFullPath(route);
        /** @type {?} */
        var text = typeof data["breadcrumbs"] === 'string'
            ? data["breadcrumbs"] : data["breadcrumbs"].text || data["text"] || path;
        text = stringFormat(text, route.data);
        /** @type {?} */
        var crumbs = [
            {
                text: text,
                path: path
            }
        ];
        return of(crumbs);
    };
    /**
     * @param {?} route
     * @return {?}
     */
    McBreadcrumbsResolver.prototype.getFullPath = /**
     * @param {?} route
     * @return {?}
     */
    function (route) {
        /** @type {?} */
        var relativePath = function (segments) {
            return segments.reduce(function (a, v) { return (a += '/' + v.path); }, '');
        };
        /** @type {?} */
        var fullPath = function (routes) {
            return routes.reduce(function (a, v) { return (a += relativePath(v.url)); }, '');
        };
        return fullPath(route.pathFromRoot);
    };
    return McBreadcrumbsResolver;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
var McBreadcrumbsService = /** @class */ (function () {
    function McBreadcrumbsService(_router, _config, _injector) {
        var _this = this;
        this._router = _router;
        this._config = _config;
        this._injector = _injector;
        this._breadcrumbs = new BehaviorSubject([]);
        this._defaultResolver = new McBreadcrumbsResolver();
        this._router.events
            .pipe(filter(function (x) { return x instanceof NavigationEnd; }))
            .subscribe(function (event) {
            /** @type {?} */
            var route = _router.routerState.snapshot.root;
            _this._resolveCrumbs(route)
                .pipe(flatMap(function (x) { return x; }), distinct(function (x) { return x.text; }), toArray(), flatMap(function (x) {
                if (_this._config.postProcess) {
                    /** @type {?} */
                    var y = _this._config.postProcess(x);
                    return wrapIntoObservable(y).pipe(first());
                }
                else {
                    return of(x);
                }
            }))
                .subscribe(function (x) {
                _this._breadcrumbs.next(x);
            });
        });
    }
    Object.defineProperty(McBreadcrumbsService.prototype, "crumbs$", {
        get: /**
         * @return {?}
         */
        function () {
            return this._breadcrumbs;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @param {?} route
     * @return {?}
     */
    McBreadcrumbsService.prototype._resolveCrumbs = /**
     * @param {?} route
     * @return {?}
     */
    function (route) {
        /** @type {?} */
        var crumbs$;
        /** @type {?} */
        var data = route.routeConfig && route.routeConfig.data;
        if (data && data["breadcrumbs"]) {
            /** @type {?} */
            var resolver = void 0;
            if (data["breadcrumbs"].prototype instanceof McBreadcrumbsResolver) {
                resolver = this._injector.get(data["breadcrumbs"]);
            }
            else {
                resolver = this._defaultResolver;
            }
            /** @type {?} */
            var result = resolver.resolve(route, this._router.routerState.snapshot);
            crumbs$ = wrapIntoObservable(result).pipe(first());
        }
        else {
            crumbs$ = of([]);
        }
        if (route.firstChild) {
            crumbs$ = crumbs$.pipe(concat(this._resolveCrumbs(route.firstChild)));
        }
        return crumbs$;
    };
    McBreadcrumbsService.decorators = [
        { type: Injectable },
    ];
    /** @nocollapse */
    McBreadcrumbsService.ctorParameters = function () { return [
        { type: Router },
        { type: McBreadcrumbsConfig },
        { type: Injector }
    ]; };
    return McBreadcrumbsService;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
var McBreadcrumbsComponent = /** @class */ (function () {
    function McBreadcrumbsComponent(service) {
        this.service = service;
        this.subscriptions = new Array();
    }
    /**
     * @return {?}
     */
    McBreadcrumbsComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        /** @type {?} */
        var s = this.service.crumbs$.subscribe(function (x) {
            _this.crumbs = x;
        });
    };
    /**
     * @return {?}
     */
    McBreadcrumbsComponent.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        this.subscriptions.forEach(function (x) { return x.unsubscribe(); });
    };
    McBreadcrumbsComponent.decorators = [
        { type: Component, args: [{
                    // tslint:disable-next-line:component-selector
                    selector: 'mc-breadcrumbs',
                    template: "<ol *ngIf=\"crumbs.length\" class=\"breadcrumb\">\n  <li *ngFor=\"let crumb of crumbs; let last = last\" [ngClass]=\"{ 'active': last }\" class=\"breadcrumb-item\">\n    <a *ngIf=\"!last\" [routerLink]=\"crumb.path\">{{ crumb.text }}</a>\n    <span *ngIf=\"last\">{{ crumb.text }}</span>\n  </li>\n</ol>\n"
                },] },
    ];
    /** @nocollapse */
    McBreadcrumbsComponent.ctorParameters = function () { return [
        { type: McBreadcrumbsService }
    ]; };
    return McBreadcrumbsComponent;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
var McBreadcrumbsModule = /** @class */ (function () {
    function McBreadcrumbsModule() {
    }
    /**
     * @return {?}
     */
    McBreadcrumbsModule.forRoot = /**
     * @return {?}
     */
    function () {
        return {
            ngModule: McBreadcrumbsModule,
            providers: [McBreadcrumbsService, McBreadcrumbsConfig]
        };
    };
    McBreadcrumbsModule.decorators = [
        { type: NgModule, args: [{
                    imports: [CommonModule, RouterModule],
                    declarations: [McBreadcrumbsComponent],
                    exports: [McBreadcrumbsComponent]
                },] },
    ];
    return McBreadcrumbsModule;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */

export { McBreadcrumbsComponent, McBreadcrumbsService, McBreadcrumbsConfig, McBreadcrumbsResolver, McBreadcrumbsModule };
